### PR TITLE
♻️ REFACTOR: package code

### PR DESCRIPTION
**Pull request by chrisjsewell**
Essentially re-writes the most of the code, to implement a number of improvements and bug fixes, but the output LaTeX is shown (via tests) to be essentially the same.

- move event functions to separate file (and rename)
- add sphinx based tests, which provide comparisons of the doctrees create with/without this extension, before/after post-transforms have been applied. This also shows how the package can be used as a "standalone" sphinx extension.
- use `importlib.resources` for accessing `jupyterBook.cls`
- Remove enforcement of `myst-parser` `amsmath` extension
- make `myst_nb` requirement optional, and check for compatibility
- convert class/function nameing from pascalCase to camel_case (and enforce with pep8-naming linter):
  - replace `getFilenameWithSubpath` with `is_root_document`
  - replace `removeExtension` with `remove_suffix`
- replace `H2Node` and `H3Node` by `RootHeader` node, which applies to any sub-section of the root file (i.e. also H4 etc)
- added `app.add_config_value("jblatex_load_imgconverter", True, "env")` to enable loading of sphinx.ext.imgconverter to be turned off
- remove reliance on _toc.yml, instead adding `app.add_config_value("jblatex_captions_to_parts", None, "env", (type(None), bool))`
  - If `None` and `app.env.external_site_map` is present (from sphinx-external-toc), then this will be used to "infer" `latex_toplevel_sectioning` and whether to convert captions to parts.
- Move all doctree restructuring to the post-transform, so that the stored doctrees are builder agnostic (as should be the case). The transform (which is now called for all builders) only adds extra node attributes for use by the post-transform.
- a lot more code documentation and added to README

### CODE REVIEW

**Comment by chrisjsewell** [comment]
Not finished, but publishing so @mmcky / @AakashGfude don;t do any conflicting work

TODO: rewrite LatexMasterDocTransforms, ToctreeTransforms, handleSubSections, add more testing

**Comment by mmcky** [comment]
@AakashGfude there has been a lot of discussion over the past few days re: `sphinx-external-toc` so let's catchup this morning on zoom and I will bring you up to speed.

File: jupyterbook_latex/events.py
Comment (chrisjsewell) [inline jupyterbook_latex/events.py:101]: This is faulty logic, which will fail for rebuilds, if you run e.g. a HTML build then a LaTeX build; the HTML build would store the doctree (without having run `LatexRootDocTransforms`) then the latex build will take that doctree and go straight to post-transforms,
i.e. you can't have transforms that are builder specific, only post-transforms.

Comment (chrisjsewell) [reply jupyterbook_latex/events.py:101]: can the functionality in this transform be moved to a post-transform

File: jupyterbook_latex/events.py
Comment (chrisjsewell) [inline jupyterbook_latex/events.py:100]: I don't think we should be enforcing this extension, people might want to use e.g. https://github.com/missinglinkelectronics/sphinxcontrib-svg2pdfconverter

File: jupyterbook_latex/events.py
Comment (chrisjsewell) [inline jupyterbook_latex/events.py:93]: I don't think we should be enforcing the `amsmath` extension

File: jupyterbook_latex/events.py
Comment (AakashGfude) [reply jupyterbook_latex/events.py:100]: @chrisjsewell `sphinx.ext.imgconverter` also handles gif images, which `sphinxcontrib-svg2pdfconverter` doesn't. Maybe we can keep this as default, and allow ppl to override it with other extension?

File: jupyterbook_latex/events.py
Comment (AakashGfude) [reply jupyterbook_latex/events.py:93]: @chrisjsewell Yeah,  I think you are right.

File: jupyterbook_latex/events.py
Comment (mmcky) [reply jupyterbook_latex/events.py:100]: I think having an opinionated set of tools that we need to deliver a base `pdf` is a good thing -- but perhaps this should be made an option with the default value being `sphinx.ext.imgconverter`

File: jupyterbook_latex/events.py
Comment (chrisjsewell) [reply jupyterbook_latex/events.py:101]: @mmcky @AakashGfude was there a specific reason for splitting the "root doc restructure" into a transform and post-transform? e.g. is there certain things you know that need to happen before/after other things?

Comment (AakashGfude) [reply jupyterbook_latex/events.py:101]: @chrisjsewell I remember there was a reason to do this in transform because of the state of the doctree required. Will go through the code to pinpoint the exact reasons.

File: jupyterbook_latex/events.py
Comment (chrisjsewell) [reply jupyterbook_latex/events.py]: removed

File: jupyterbook_latex/events.py
Comment (chrisjsewell) [reply jupyterbook_latex/events.py]: I've moved all the restructuring to the post-transform, which works

File: jupyterbook_latex/events.py
Comment (chrisjsewell) [reply jupyterbook_latex/events.py]: added `app.add_config_value("jblatex_load_imgconverter", True, "env")`

File: tests/test_jupyter_book/test_toc.tex
Comment (chrisjsewell) [inline tests/test_jupyter_book/test_toc.tex:20]: The current jupyter-book creates this weird list if a `tableofcontents` directive is used.
But this is no longer there when moving to sphinx-external-toc, which I believe is correct.
I've already noted this in the jupyter-book PR: https://github.com/executablebooks/jupyter-book/pull/1293#discussion_r615433968

**Comment by chrisjsewell** [comment]
ok this is ready to review

**Comment by mmcky** [comment]
thanks @chrisjsewell -- we will review this today.

File: tests/test_jupyter_book/test_toc.tex
Comment (AakashGfude) [reply tests/test_jupyter_book/test_toc.tex:20]: @chrisjsewell will add `tableofcontents` directive logic here, after this PR gets merged.

Comment (chrisjsewell) [reply tests/test_jupyter_book/test_toc.tex:20]: but I don't understand why this list should be there in the first place? Is this something you want?

File: setup.cfg
Comment (mmcky) [inline setup.cfg:45]: @AakashGfude this PR should be reviewed against this `version` of `sphinx-external-toc`

File: tests/test_jupyter_book/test_toc.tex
Comment (AakashGfude) [reply tests/test_jupyter_book/test_toc.tex:20]: @chrisjsewell because `tableofcontents` directive was transformed into a list for latex. using the `toctree` directive directly in latex(like we do for HTML) includes the files themselves instead of just returning a list of file titles.

File: jupyterbook_latex/nodes.py
Comment (AakashGfude) [inline jupyterbook_latex/nodes.py:57]: Thank you for transforming it to a single node, was planning this in a later release. I think we also need to accommodate sizes apart from these two, like H4, in case people have subsubsections, etc.

File: tests/test_jupyter_book/test_toc.tex
Comment (AakashGfude) [reply tests/test_jupyter_book/test_toc.tex:20]: I tried using `\tableofcontents` command, or packages like [minitoc](https://www.ctan.org/pkg/minitoc#:~:text=The%20minitoc%20package%20allows%20you,every%20chapter%2C%20part%20or%20section.&text=If%20the%20type%20of%20document,level%20secttocs%2C%20sectlofs%20and%20sectlots.) , but all of them had issues. For example, they specified the page numbers as well of entries, something like:
Screen Shot 2021-04-23 at 2 17 51 pm
 
But, we wanted a simpler version of it(like that for in HTML), only having the title names, like:

Screen Shot 2021-04-23 at 2 20 59 pm

 and directly using `toctree` directive did not help(for the reason mentioned in the comment before), so created a bullet list by myself.

Comment (mmcky) [reply tests/test_jupyter_book/test_toc.tex:20]: @AakashGfude I do like the simplicity of these `contents` listings. Perhaps we should keep this as is and iterate to see if we can find a more `LaTeX` ecosystem based approach to solving this problem.

File: MANIFEST.in
Comment (AakashGfude) [inline MANIFEST.in:17]: @chrisjsewell just wanted to know the thinking behind this. in some places we use * like https://github.com/executablebooks/jupyter-book/blob/master/MANIFEST.in#L29, and in some don't. My thinking was that in the future we might have some `.sty` files and other theme-specific files in this folder, and the * declaration will make it future-proof.

**Comment by chrisjsewell** [comment]
You shouldn't be building your own listings though, there latex packages for that

File: MANIFEST.in
Comment (chrisjsewell) [reply MANIFEST.in:17]: The `check-manifest` pre-commit hook already keeps it future proof

Comment (AakashGfude) [reply MANIFEST.in:17]: awesome.

**Review by mmcky — COMMENTED** [review]
Thanks @chrisjsewell on your work on this refactor. It is a big improvement. We will apply some of these changes (in style etc.) to `sphinx-multitoc-numbering` as well. 

I re-ran the full tests of building our lecture sites for both `html` and `pdf` (using this refactor):

- [x] https://github.com/QuantEcon/lecture-python-programming.myst/pull/135 generates the `pdf` version nicely
- [ ] https://github.com/QuantEcon/lecture-python.myst/pull/133 fails to build the `pdf` giving the following `extension` error. It seems to be related to `myst_nb`?

```bash
Extension error:
Handler <function validate_config_values at 0x7fb649c8e550> for event 'config-inited' threw an exception (exception: entry_points() got an unexpected keyword argument 'group')
Traceback (most recent call last):
  File "/usr/share/miniconda3/envs/quantecon/lib/python3.8/site-packages/sphinx/events.py", line 110, in emit
    results.append(listener.handler(self.app, *args))
  File "/usr/share/miniconda3/envs/quantecon/lib/python3.8/site-packages/myst_nb/__init__.py", line 256, in validate_config_values
    load_renderer(app.config["nb_render_plugin"])
  File "/usr/share/miniconda3/envs/quantecon/lib/python3.8/site-packages/myst_nb/render_outputs.py", line 80, in load_renderer
    eps = entry_points(group="myst_nb.mime_render", name=name)
TypeError: entry_points() got an unexpected keyword argument 'group'
```

the [build logs are here](https://github.com/QuantEcon/lecture-python.myst/pull/133/checks?check_run_id=2434074806) and the artifact is a complete `_build` folder.

**Comment by mmcky** [comment]
thanks @chrisjsewell I confirm both of our current `project` tests cases are now building via `CI`

File: jupyterbook_latex/nodes.py
Comment (mmcky) [reply jupyterbook_latex/nodes.py:95]: @AakashGfude is this still a concern? Perhaps we can document the limitation for now. Can you coordinate this?

**Review by mmcky — APPROVED** [review]
@chrisjsewell nice work!

File: jupyterbook_latex/nodes.py
Comment (AakashGfude) [reply jupyterbook_latex/nodes.py:95]: @mmcky it has mimicked the previous implementation in this package. So, I think it is good for this PR. We should create an issue, and create a new PR for it.

**Review by AakashGfude — APPROVED** [review]
Awesome work @chrisjsewell . The only thing left is to transfer `tableofcontents` directive here, for which I will create an issue and PR. Thanks so much.

**Comment by chrisjsewell** [comment]
Thanks, I'm not sure what you mean by transfer the `tableofcontents` directive though, that's part of `sphinx-external-toc` now, you wouldn't use it here. If you want those bullet lists of contents (which should be optional, with "pure" latex alternatives also made available) you'd just be looking for non-hidden toctrees

**Comment by chrisjsewell** [comment]
Also, just to clarify, I'll merge this once I've finished the final point

**Comment by AakashGfude** [comment]
> Thanks, I'm not sure what you mean by transfer the `tableofcontents` directive though, that's part of `sphinx-external-toc` now, you wouldn't use it here. If you want those bullet lists of contents (which should be optional, with "pure" latex alternatives also made available) you'd just be looking for non-hidden toctrees

ooh. I could not see the `tableofcontents` directive result in the pdf, so assumed it was not handled for latex yet. 

A screenshot:
Screen Shot 2021-04-27 at 11 29 36 am

[book.pdf](https://github.com/executablebooks/jupyterbook-latex/files/6381277/book.pdf)

**Comment by chrisjsewell** [comment]
The `tableofcontents` now works no different to if you were just using a regular `toctree` (without sphinx-external-toc). 
So the point is you should be thinking of a "generic" solution that would work with any regular sphinx build (as the rest of the package now does)

**Comment by AakashGfude** [comment]
> The `tableofcontents` now works no different to if you were just using a regular `toctree` (without sphinx-external-toc).
> So the point is you should be thinking of a "generic" solution that would work with any regular sphinx build (as the rest of the package now does)

Sorry, maybe I did not make my point clear earlier. The problem I faced when using a regular `toctree` with `tableofcontents` like the rest of the sphinx build, was that in the case of latex, it literally included the files itself, instead of just links to the file, as was expected. If I wanted to change this behavior, I had to interfere with the sphinx latex writer.

So, while the rest of the builders were using the inline `toctree` directive of that page, the latex builder needed special handling. I am very much looking forward to any generic solutions you have for the latex builder. But, didn't see any solution in the `sphinx-external-toc` PR, so was enquiring.

**Comment by mmcky** [comment]
@AakashGfude if we need special handling of the `toc` for the `latex` builder then this package is the best place for that. 

> I had to interfere with the sphinx latex writer.

Do we need to adjust the `writer` or the `translator`? 
Could we just override the method that is producing output that we don't want (that has the filenames included)?

To assist with this discussion could you please add some screenshots showing the `pdf` output that `sphinx` produces. I remember seeing it (and it didn't look good) and that's when we went with the special handling for displaying `local contents`.

**Comment by chrisjsewell** [comment]
> I think the best way forward is to use the simple lists that @AakashGfude has put together and we can open an issue to replace it with a more pure latex implementation using some latex package to generate simple local contents listings. @AakashGfude can you open a PR to add in this support.

this should not be that difficult:

1. The point of this PR, and thus any future work, is that for builder specific extensions you can "gather" information in a transform, but you **cannot** implement any modifications of the AST until a post-transform.
2. The list addition must be optional, and should be one option of how to generate a toc.

**Pull request merged**
